components/dashboard/src/workspaces/CreateWorkspacePage.tsx

```diff
@@ -0,0 +1,346 @@
+        props.ideSettings?.useLatestVersion !== undefined
+            ? props.ideSettings.useLatestVersion
+            : !!user?.additionalData?.ideSettings?.useLatestVersion,
+    );
+    const [selectedIde, setSelectedIde] = useState(
+        props.ideSettings?.defaultIde !== undefined
+            ? props.ideSettings.defaultIde
+            : user?.additionalData?.ideSettings?.defaultIde,
+    );
+    const [selectedWsClass, setSelectedWsClass] = useState<string | undefined>(props.workspaceClass);
+    const [errorWsClass, setErrorWsClass] = useState<string | undefined>(undefined);
+    const [repo, setRepo] = useState<string | undefined>(location.hash.substring(1));
+    const onSelectEditorChange = useCallback(
+        (ide: string, useLatest: boolean) => {
+            setSelectedIde(ide);
+            setUseLatestIde(useLatest);
+        },
+        [setSelectedIde, setUseLatestIde],
+    );
+    const [errorIde, setErrorIde] = useState<string | undefined>(undefined);
+    const [creating, setCreating] = useState<boolean>(false);
+
+    const [createResult, setCreateResult] = useState<WorkspaceCreationResult | undefined>(undefined);
+    const [createError, setCreateError] = useState<StartWorkspaceError | undefined>(undefined);
+    const [selectAccountError, setSelectAccountError] = useState<SelectAccountPayload | undefined>(undefined);
+
+    const createWorkspace = useCallback(
+        async (options?: Omit<GitpodServer.CreateWorkspaceOptions, "contextUrl">) => {
+            // add options from search params
+            const opts = options || {};
+
+            if (!opts.workspaceClass) {
+                opts.workspaceClass = selectedWsClass;
+            }
+            if (!opts.ideSettings) {
+                opts.ideSettings = {
+                    defaultIde: selectedIde,
+                    useLatestVersion: useLatestIde,
+                };
+            }
+            if (!repo) {
+                return;
+            }
+
+            try {
+                setCreating(true);
+                const result = await getGitpodService().server.createWorkspace({
+                    contextUrl: repo,
+                    ...opts,
+                });
+
+                setCreateResult(result);
+            } catch (error) {
+                setCreateError(error);
+            } finally {
+                setCreating(false);
+            }
+        },
+        [repo, selectedIde, selectedWsClass, useLatestIde],
+    );
+
+    const onClickCreate = useCallback(() => createWorkspace(), [createWorkspace]);
+
+    if (SelectAccountPayload.is(selectAccountError)) {
+        return (
+            <SelectAccountModal
+                {...selectAccountError}
+                close={() => {
+                    window.location.href = gitpodHostUrl.asAccessControl().toString();
+                }}
+            />
+        );
+    }
+    if (createResult?.workspaceURL) {
```

Comment:
Could we shift this out of the render loop and into the `createWorkspace` function right after we get the `result` from the api call? It would be much more explicit that we're navigating after creation if there's a workspaceUrl on the result. It would also remove the need to have component state for that result.

Reply:
Not sure I understood your suggestion, the navigation only happens on close of the Modal which needs to be rendered.

Reply:
`createWorkspace` could look like something like this, and use a more explicit imperative approach that doesn't have the indirection of setting a result as state, and expecting a re-render of the component to then trigger a navigation. I was wrong about the `ExistingWorkspaceModal` though, didn't realize it needs the result to render, but we could still make it a bit more clear I think:

```ts
// In createWorkspace()
const result = await getGitpodService().server.createWorkspace({
    contextUrl: repo,
    organizationId: team?.id,
    ...opts,
});

// handle navigation here
if(result?.workspaceURL) {
    window.location.href = result.workspaceURL;
    return;
}

// and here
if (result?.createdWorkspaceId) {
    history.replace(`/start/${createResult.createdWorkspaceId}`);
    return;
}

// This one is debatable, I didn't realize the modal render needs the result, but this could be more clear by something like...
if (result?.existingWorkspaces) {
    setExistingWorkspaceResult(result)
    return;
}

//... then in the rendering
if (existingWorkspaceResult) {
    return (
        <ExistingWorkspaceModal 
            result={existingWorkspaceResult} 
            createWorkspace={createWorkspace} onClose={() => { }} 
        />
    )
}
```